src/dataset/create_dataset.py

```diff
@@ -19,240 +20,141 @@
-                    blocks = text_splitter.split(text, signature)
-                    self.embedding_strings.extend(blocks)
-                    self.embeddings_metadata_index.extend([self.total_articles_count-1] * len(blocks))
-                    
-                    # Update counts
-                    self.total_char_count += len(text)
-                    self.total_word_count += len(text.split())
-                    self.total_sentence_count += len(split_into_sentences(text))
-                    self.total_block_count += len(blocks)
-                    
-                except MissingDataException as e:
-                    if str(e) not in error_count_dict:
-                        error_count_dict[str(e)] = 0
-                    error_count_dict[str(e)] += 1
+
+        # Load the dataset. streaming allows you to load one entry at a time, 
+        # so entries can be processed before the entire dataset has been saved.
+        # iterable_data = load_dataset('StampyAI/alignment-research-dataset', 'aisafety.info', split='train', streaming=True)
+
+        iterable_data = load_dataset('StampyAI/alignment-research-dataset', 'all', split='train', streaming=True)
+        
+        for entry in tqdm(iterable_data):            
+            """Checks"""
+
+            # if we specified custom sources, only include entries from those sources
+            if (self.custom_sources is not None) and (entry['source'] not in self.custom_sources):
+                continue
+
+            # raise error if the entry does not contain the required metadata
+            self.contains_required_metadata(entry)
+
+            #if the text is too short, ignore this text
+            if len(entry['text']) < 500:
+                continue
+
+
+            """Checks are done, so we construct the metadata."""
+
+            # Get id, source, title, text, url, date_published, authors, and summary
+            entry_id: str = entry['id']
+            source: str = entry['source']
+            title: str = entry['title']
+            text: str = entry['text']
+            url: str = entry['url']
+            date_published: str = entry['date_published']
+            authors: list = entry['authors']
+            # summary is ignored for now, since most sources lack one. TODO: add summary. see self.metadata code as well.
+            
+            # Get signature
+            if authors:
+                signature = f"Title: {title}, Authors: {get_authors_str(authors)}"
+            else:
+                signature = f"Title: {title}"
+
+            # We use the text_splitter to get the chunks from the entry, 
+            # and we add a metadata element for each new chunk we add to the dataset.
+            chunks = text_splitter.split(text, signature)
+            num_chunks = len(chunks)
+            
+            for i in range(num_chunks):
+                self.metadata.append({
```

Comment:
this could be a potential problem on machines without much memory - it might be better to change this method into a a generator. Though that would mess up the `save_data` function...

Reply:
Are you saying we essentially don't want to have to hold into memory the entire content of the dataset, which is what we do by constructing self.metadata? If so, yeah I agree, if we add the information from the entries into the sqlite3 database that removes the necessity of storing self.metadata at all, correct? 

Another thing this reminds me of is that I may not understand how the streaming on the load_dataset() function works, because my understanding is that, as we go through the iterable, we load more and more bits of the dataset, but at the end, we hold the entire dataset into memory, and maybe we crash if that's too big for our memory - so wouldn't we want to go through the dataset but forget about previously seen entries? Maybe that's what it does, I'm not sure

Reply:
Unless I'm misunderstanding... Since you are already rate-limited to 3500 embeddings per minute, there may not be a benefit to have everything in memory all at once. Personally, I didn't run into memory problems running in the colab notebook and SentenceTransformer embeddings were able to batch it properly but OepnAI may be different. 

However, I do think we should still store most of the key metadata (id, source, title, authors, date_published, chunked_text but can skip full_text and summary for now) in pinecone as well. Sqlite3 should only be used to keep track of data for ingesting ARD, chunking, and embedding. Whatever the chatbot needs to query & generate answers should be stored in pinecone to reduce an extra dependencies (risk of failure) at runtime. The metadata feature of pinecone is supposed to be one of its benefits over other vector DBs.

Reply:
Although... it might be nice to have a counter in sqlite to keep track of which entries in the ARD are used as sources most frequently. Anyway, once the database is online, there's a bunch of other things we would like to track. Not sure sqlite can handle the load but that's a separate discussion.

Reply:
I definitely agree we don't want everything in memory at once. It's good that it works for the colab but it's still practical if my brother and I can test it without those concerns, but also it's not top priority I guess as long as it works. And alright for pinecone, since lw has less entries than before, storing those metadata values should be OK. So I guess in that case the DB is only used for constructing the pinecone, or when we want to get information from keys that aren't in the 8 basic ones, like tags or "book_title" and so on?